### PR TITLE
feat(tier4_autoware_utils): add published time debug class into utils

### PR DESCRIPTION
**Pull request by brkay54**
## Description

Part of:
- https://github.com/autowarefoundation/autoware.universe/issues/6255

Required for:
- https://github.com/autowarefoundation/autoware.universe/pull/6490

Depends on:
- https://github.com/autowarefoundation/autoware/pull/4212
- https://github.com/autowarefoundation/autoware_internal_msgs/pull/1

This interface would be enabled only if the use_published_time = true. 

## Tests performed

Tested on PSim and e2e_simulator, worked well.

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp
Comment (mitsudome-r) [inline common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp:38]: We can add condition check for subscriber count at the beginning of this function and do nothing if there are no subscribers.

File: common/tier4_autoware_utils/package.xml
Comment (yukkysaito) [inline common/tier4_autoware_utils/package.xml:19]: related link : https://github.com/autowarefoundation/autoware_msgs/pull/83#issuecomment-1963435608

File: common/tier4_autoware_utils/package.xml
Comment (yukkysaito) [inline common/tier4_autoware_utils/package.xml:18]: Is this necessary? :pray:

File: common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp
Comment (yukkysaito) [inline common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp:1]: This one might be better.
```suggestion
// Copyright 2024 The Autoware Contributors
```

File: common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp
Comment (yukkysaito) [inline common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp:52]: https://github.com/autowarefoundation/autoware-documentation/blob/70a2509315371e9378e7383246ea565ae35ebbe7/docs/contributing/coding-guidelines/ros-nodes/topic-namespaces.md?plain=1#L83

```suggestion
    rclcpp::Node * node, const std::string & end_name = "~/debug/published_time",
```

**Review by yukkysaito — APPROVED** [review]
LGTM.
I have not tested it myself. Please make sure that the tutorial is working before merging. :pray:

**Comment by brkay54** [comment]
Hi @xmfcx, it is ready for review. Also, I updated the packages we will add this debug publisher :
- https://github.com/autowarefoundation/autoware.universe/pull/6490
The readme of the packages was updated. Both packages and `PublishedTime` publisher debug class are ready for review.

File: common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp
Comment (xmfcx) [inline common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp:31]: What is Gid?

File: common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp
Comment (xmfcx) [inline common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp:62]: We don't need a way to disable this with declared params, you can remove this `use_published_time` param and its logic.

Could you remove this create function and make it a constuctor instead?

File: common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp
Comment (xmfcx) [inline common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp:126]: I've analyzed the overall implementation and it is really well thought out.

However, this overloaded hash and equality operators make the code hard to read.

Suggestion:

use a simple unordered map like:
```cpp
std::unordered_map<std::string, rclcpp::Publisher<PublishedTime>::SharedPtr> publishers_;
```

have this function:
```cpp
std::string gid_to_str_key(const rmw_gid_t & gid) const {
  return std::string(reinterpret_cast<const char *>(gid.data), RMW_GID_STORAGE_SIZE);
}
```

and whoever wants to access the map uses this step:
```
const std::string gid_key = gid_to_str_key(publisher->get_gid());
publishers_.find(gid_key) ...
publishers_[gid_key] = ...
```

This makes it overall more readable.

**Comment by xmfcx** [comment]
Overall nice class, from the usage of the class, my suggestions only remove the `create` function and promotes usage of the constructor instead.

File: common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp
Comment (brkay54) [reply common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp]: Makes sense, I removed.

**Comment by brkay54** [comment]
@xmfcx Thank you for the review! I updated the branch w.r.t your suggestions. Could you re-check?

**Comment by xmfcx** [comment]
Could you move `struct GidCompare` to the class, as a private member, closer to its usage?

Reasons:
1. Encapsulation is improved by moving `GidCompare` inside `PublishedTimePublisher`, making it clear that this comparison logic is specific to the class's internal workings.
2. Cohesion is enhanced by keeping related functionalities together, making the class more self-contained and understandable.

Example:
```cpp
private:
  rclcpp::Node * node_;
  std::string end_name_;
  rclcpp::QoS qos_;

  struct GidCompare
  {
    bool operator()(const rmw_gid_t & lhs, const rmw_gid_t & rhs) const
    {
      return std::memcmp(lhs.data, rhs.data, RMW_GID_STORAGE_SIZE) < 0;
    }
  };

  std::map<rmw_gid_t, rclcpp::Publisher<PublishedTime>::SharedPtr, GidCompare> publishers_;

  void ensure_publisher_exists(const rmw_gid_t & gid_key, const std::string & topic_name);
```

**Comment by xmfcx** [comment]
Placing the `using` directive for `PublishedTime` outside the class at the namespace level makes it globally accessible within the namespace, which could lead to naming conflicts with other entities. This broader scope unnecessarily exposes the alias, potentially complicating the codebase and affecting readability for developers not directly working with the `PublishedTimePublisher` class.

Moving the `using` directive for `PublishedTime` inside the `PublishedTimePublisher` class confines its scope to where it is actually used, enhancing code readability and maintainability. This approach avoids potential naming conflicts and makes the class definition more self-contained and easier to understand.

```cpp
private:
  rclcpp::Node * node_;
  std::string end_name_;
  rclcpp::QoS qos_;

  using PublishedTime = autoware_internal_msgs::msg::PublishedTime;
```

File: common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp
Comment (xmfcx) [inline common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp:78]:
```suggestion

    // if the publisher is not in the map, create a new publisher for published time
    ensure_publisher_exists(gid_key, publisher->get_topic_name());
```

We can remove the topic_name variable since it is used only here.

File: common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp
Comment (xmfcx) [inline common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp:56]:
```suggestion

    // if the publisher is not in the map, create a new publisher for published time
    ensure_publisher_exists(gid_key, publisher->get_topic_name());
```

same as https://github.com/autowarefoundation/autoware.universe/pull/6440/files#r1519499450

File: common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp
Comment (xmfcx) [inline common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp:95]:
```suggestion
  std::string publisher_topic_suffix_;
```

**Review by xmfcx — APPROVED** [review]
thanks!

**Comment by xmfcx** [comment]
https://github.com/autowarefoundation/autoware.universe/actions/runs/8245809015/job/22550518856?pr=6440#step:7:99
fails

**Comment by xmfcx** [comment]
https://github.com/autowarefoundation/autoware.universe/pull/6440/commits/a1a595c6e509a61088030592a2e527c6d359f50b should fix it.

**Comment by brkay54** [comment]
@soblin @TakaHoribe @takayuki5168 Hi, we need a code owner approvement. Could you check the PR?

**Review by takayuki5168 — CHANGES_REQUESTED** [review]
@brkay54 
Thank you for the contribution. Can you create a unit test for the feature? Generally, we create a test for functions in the tier4_autoware_utils package.

**Comment by xmfcx** [comment]
@brkay54 could you open up a new PR for the tests of the code added here? I will merge it, it is blocking some other PRs.

**Pull request merged**
